src/liballoc/sync.rs

```diff
@@ -866,12 +866,10 @@ impl<T: ?Sized> Arc<T> {
     unsafe fn drop_slow(&mut self) {
         // Destroy the data at this time, even though we may not free the box
         // allocation itself (there may still be weak pointers lying around).
-        ptr::drop_in_place(&mut self.ptr.as_mut().data);
+        ptr::drop_in_place(Self::get_mut_unchecked(self));
 
-        if self.inner().weak.fetch_sub(1, Release) == 1 {
-            acquire!(self.inner().weak);
-            Global.dealloc(self.ptr.cast(), Layout::for_value(self.ptr.as_ref()))
-        }
+        // Convert ourselves into a weak pointer and drop it
+        Weak { ptr: self.ptr };
```

Comment:
Uh, is this change needed? Feels mostly orthogonal. I'd prefer to keep such tricky PRs as small and focused as possible.

Reply:
I suppose not, but it does de-duplicate some fairly tricky logic, and with the change to Weak::drop that this PR makes, these two implementations would otherwise deviate. If you feel strongly I can revert it, but it seems a little silly.

Reply:
Hm, makes sense. You now have an extra `if` here but this is already called `drop_slow` so it's probably not a big deal.

Two comments:
* "convert ourselves into weak" sounds like a `fn(self) -> Weak` kind of situation, but that's not what this is. So I'd rather say "drop the weak count held by all strong references" or so.
* When dropping is the core part of what happens, I think it is better to write `drop(Weak { ... });`, to make that clear and to also make it so that one can see clearly where the drop happens.

Reply:
- Updated the comment.
- Yep, also I would expect this to check to be trivially optimised out - our previous access to `ptr` implicitly asserts that it's a valid pointer.